nomad/event_sink_manager.go

```diff
@@ -402,148 +379,134 @@ func (m *ManagedSink) Run() error {
 // should be reloaded.
-func (m *ManagedSink) ResetWatchSet() bool {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+func (m *ManagedSink) resetSink() {
+	currentSubCancel := m.sinkCancelFn
 
+	// Fetch the Event Sink
 	ws := m.stateFn().NewWatchSet()
 	sink, err := m.stateFn().EventSinkByID(ws, m.Sink.ID)
 	if err != nil {
-		return true
+		// Log the error, continue to set the sink to nil
+		m.logger.Error("error querying for latest event sink", "sink", m.Sink.ID, "error", err)
 	}
 
-	if sink == nil {
-		return true
-	}
-
-	// Update the watchCh
-	m.watchCh = ws.WatchCh(m.sinkCtx)
+	// Set our Sink to the new one
+	oldSink := m.Sink
+	m.Sink = sink
+	m.watchCh = ws.WatchCh(m.stopCtx)
 
-	current := m.Sink
-
-	// Only reload if something that affects the sink destination
-	// or subscription
-	if current.Address != sink.Address ||
-		current.Type != sink.Type ||
-		!reflect.DeepEqual(current.Topics, sink.Topics) {
-		return true
+	// If the new Sink value is nil or if the sink has changed in a meaningful
+	// way cancel the current subscription.
+	if m.Sink == nil || m.Sink.Changed(oldSink) {
+		// Cancel the existing sinkCtx since we need to reload the subscription
+		// and sink writer
+		currentSubCancel()
+		// Set the new sinkCtx and cancelFn
+		sinkCtx, cancel := context.WithCancel(m.stopCtx)
+		m.sinkCtx = sinkCtx
+		m.sinkCancelFn = cancel
 	}
 
-	return false
 }
 
-// Reload reloads and resets a ManagedSink.
-func (m *ManagedSink) Reload() error {
-	// Exit if shutting down
-	if err := m.stopCtx.Err(); err != nil {
-		return err
-	}
-
-	// Unsubscribe incase we haven't yet
-	m.Subscription.Unsubscribe()
-
-	// Fetch our updated or changed event sink with a new watchset
-	ws := memdb.NewWatchSet()
-	ws.Add(m.stateFn().AbandonCh())
-	sink, err := m.stateFn().EventSinkByID(ws, m.Sink.ID)
-	if err != nil {
-		return err
-	}
-
+// subscribe starts a new subscription to send to the SinkWriter
+func (m *ManagedSink) subscribe() error {
 	// Sink has been deleted, stop
-	if sink == nil {
+	if m.Sink == nil {
 		return ErrEventSinkDeregistered
 	}
 
-	// Reconfigure the sink writer
-	writer, err := stream.NewWebhookSink(sink)
+	// Unsubscribe from the current subscription if there is one
+	if m.subscription != nil {
+		m.subscription.Unsubscribe()
+	}
+
+	// Generate the sink writer
+	writer, err := stream.NewWebhookSink(m.Sink)
 	if err != nil {
 		return fmt.Errorf("generating sink writer for sink %w", err)
 	}
+	m.sinkWriter = writer
 
-	// Reset values we are updating
-	sinkCtx, cancel := context.WithCancel(m.stopCtx)
-	m.sinkCtx = sinkCtx
-	m.cancelFn = cancel
-	m.SinkWriter = writer
-	m.Sink = *sink
-	m.watchCh = ws.WatchCh(sinkCtx)
+	// Set the starting index for the new subscription. The locally tracked
+	// Index may be ahead of the one periodically sent to raft, use whichever
+	// is greater to reduce duplicates
+	startIndex := m.Sink.LatestIndex
+	localIndex := atomic.LoadUint64(&m.lastSuccess)
```

Comment:
It's more obvious to me now in this PR, but I really like your use of atomics to avoid a bunch of awkward lock/unlock code throughout.